Classes/Views/GutterLayoutManager.swift

```diff
@@ -0,0 +1,256 @@
+        let maxLength = "\(numberOfLines)".count
+        let fillerString = String(repeating: "9", count: maxLength)
+        let maxGutterSize = (fillerString as NSString).size(withAttributes: [NSAttributedStringKey.font: gutterLayoutManager.gutterFont])
+        return ceil(maxGutterSize.width)
+    }
+}
+
+class GutterTextContainer: NSTextContainer {
+
+    override func lineFragmentRect(forProposedRect proposedRect: CGRect, at characterIndex: Int, writingDirection baseWritingDirection: NSWritingDirection, remaining remainingRect: UnsafeMutablePointer<CGRect>?) -> CGRect {
+
+        var lineFragmentRect = super.lineFragmentRect(
+            forProposedRect: proposedRect,
+            at: characterIndex,
+            writingDirection: baseWritingDirection,
+            remaining: remainingRect
+        )
+
+        guard let layoutManager = self.layoutManager as? GutterLayoutManager else { return lineFragmentRect }
+        let fragmentInsets = layoutManager.insetsForLineStarting(at: characterIndex)
+        lineFragmentRect.size.width -= fragmentInsets.left + fragmentInsets.right
+        return lineFragmentRect
+    }
+}
+
+/**
+ GutterLayoutManager allows the drawing of gutter with line number on the left side of
+ a text view. A `GutterLayoutManagerDelegate` is availabla which is being used to query the
+ actual line number to be rendered on the screen. This useful when the to be rendered line number
+ differs from the internal line number in the text.
+
+ In the code base this is used to render the correct line numbers for a diff
+ **/
+class GutterLayoutManager: NSLayoutManager {
+
+    struct Constants {
+        static let paragraphPadding: CGFloat = 4
+        static let standardGutterWidth: CGFloat = 32
+    }
+
+    private var lastParagraphLocation: Int = 0
+    private var lastParagraphNumber: Int = 0
+
+    public var showGutter: Bool = false {
+        didSet {
+            guard let textStorage = self.textStorage else { return }
+            self.invalidateDisplay(forCharacterRange: NSMakeRange(0, textStorage.length))
+        }
+    }
+
+    public var showLineNumbers: Bool = true {
+        didSet {
+            guard let textStorage = self.textStorage else { return }
+            self.invalidateDisplay(forCharacterRange: NSMakeRange(0, textStorage.length))
+        }
+    }
+
+    public var gutterTextColor = UIColor.fromHex("#A4A4A4") {
+        didSet {
+            guard let textStorage = self.textStorage else { return }
+            self.invalidateDisplay(forCharacterRange: NSMakeRange(0, textStorage.length))
+        }
+    }
+
+    public var gutterFont = UIFont.systemFont(ofSize: 11.0) {
+        didSet {
+            guard let textStorage = self.textStorage else { return }
+            self.invalidateDisplay(forCharacterRange: NSMakeRange(0, textStorage.length))
+        }
+    }
+    public var gutterDelegate: GutterLayoutManagerDelegate?
```

Comment:
If we can, maybe make the protocol require a class (`: class`) and make this weak to prevent against any retain cycles?